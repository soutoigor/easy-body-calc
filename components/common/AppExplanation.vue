<template>
  <article>
    <h1 class="explanation-title">
      <slot name="title" />
    </h1>
    <div class="explanation-content">
      <slot name="content" />
    </div>
  </article>
</template>

<style scoped>
.explanation-title {
  @apply text-5xl lg:text-7xl italic font-black;
}

.explanation-content {
  @apply text-xl flex flex-col gap-3 mt-4;
}
</style>
