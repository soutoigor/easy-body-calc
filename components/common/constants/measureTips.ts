import height from '@/assets/images/height.jpg'
import weight from '@/assets/images/weight.jpg'
import hip from '@/assets/images/hip.jpg'
import neck from '@/assets/images/neck.jpg'
import abdomen from '@/assets/images/abdomen.jpg'

export const measureTips = {
  HEIGHT: {
    TITLE: 'Height',
    IMAGE: height,
    DESCRIPTION: `
      With a measuring tape or a ruler, measure your height in centimeters,
      keep your legs together, your body straight and your head centralized.
    `,
  },
  WEIGHT: {
    TITLE: 'Weight',
    IMAGE: weight,
    DESCRIPTION: `
      Check your weight with a scale, preferably in fasting and with underwear.
    `,
  },
  HIP: {
    TITLE: 'Hip',
    IMAGE: hip,
    DESCRIPTION: `
      With a measuring tape, contour the biggest part of your hip and get the centimeters.
    `,
  },
  NECK: {
    TITLE: 'Neck',
    IMAGE: neck,
    DESCRIPTION: `
      With a measuring tape, contour your neck like a necktie, above the trapezius,
      but careful to not take the trapezius in consideration.
    `,
  },
  ABDOMEN: {
    TITLE: 'Abdomen',
    IMAGE: abdomen,
    DESCRIPTION: `
      Exhale all the air, but don't force the abdomen. And with a measuring tape, contour
      your waist right above your belly.
    `,
  },
}
