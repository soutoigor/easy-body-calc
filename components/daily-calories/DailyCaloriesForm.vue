<template>
  <form class="bodyfat-form" @submit.prevent="sendResult">
    <InputNumber
      v-model="form.height"
      data-testid="height"
      label="Height (cm)"
    />
    <InputNumber
      v-model="form.weight"
      label="Weight (cm)"
      data-testid="weight"
    />
    <InputNumber v-model="form.age" data-testid="age" label="Age" />
    <InputSelect
      v-model="form.biologicalSex"
      label="Biological sex"
      :items="biologicalSexItems"
    />
    <InputSelect
      v-model="form.activity"
      label="Exercise level"
      :items="activityLevelItems"
    />
    <AppButton type="submit" data-testid="submit" :disabled="disabled">
      Calculate
    </AppButton>
  </form>
</template>

<script lang="ts">
import InputNumber from '@/components/common/InputNumber.vue'
import InputSelect, { Items } from '@/components/common/InputSelect.vue'
import AppButton from '@/components/common/AppButton.vue'
import { computed, defineComponent, reactive } from '@nuxtjs/composition-api'
import { DailyCalories, Bmr } from './interface'

const biologicalSexItems: Items = [
  {
    label: 'Male',
    value: 'M',
  },
  {
    label: 'Female',
    value: 'F',
  },
]

const activityLevelItems: Items = [
  {
    label: 'Sedentary',
    value: 1.2,
  },
  {
    label: 'Slightly active',
    value: 1.375,
  },
  {
    label: 'Moderately active',
    value: 1.55,
  },
  {
    label: 'Active',
    value: 1.725,
  },
  {
    label: 'Very active',
    value: 1.9,
  },
]

export default defineComponent({
  components: {
    InputNumber,
    AppButton,
    InputSelect,
  },
  emits: ['result'],
  setup(_, { emit }) {
    const form = reactive<DailyCalories>({
      height: 0,
      weight: 0,
      age: 0,
      biologicalSex: null,
      activity: 0,
    })

    const disabled = computed<boolean>(() =>
      Object.values(form).some((value) => !value)
    )

    const isFemale = () => form.biologicalSex === 'F'

    const getMenBmr = ({ height, weight, age }: Bmr) =>
      88.362 + 13.397 * weight + 4.799 * height - 5.677 * age

    const getWomenBmr = ({ weight, height, age }: Bmr) =>
      447.593 + 9.247 * weight + 3.098 * height - 4.33 * age

    const calculateDailyCalories = (form: DailyCalories) => {
      const bmr = isFemale() ? getWomenBmr(form) : getMenBmr(form)
      const calories = bmr * form.activity

      return { bmr, calories }
    }

    const sendResult = () => {
      const result = calculateDailyCalories(form)

      emit('result', result)
    }

    return {
      sendResult,
      form,
      disabled,
      isFemale,
      activityLevelItems,
      biologicalSexItems,
    }
  },
})
</script>

<style scoped>
.bodyfat-form {
  @apply flex flex-col gap-8;
}
</style>
