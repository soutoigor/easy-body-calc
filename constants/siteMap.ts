export const SITE_MAP = {
  BMI: {
    label: 'BMI rate',
    url: '/bmi',
  },
  BODY_FAT: {
    label: 'Body fat',
    url: '/body-fat',
  },
  MACROS: {
    label: 'Daily macros',
    url: '/daily-macros',
  },
  CALORIES: {
    label: 'Daily calories',
    url: '/bmr-daily-calories',
  },
  GENETIC_POTENTIAL: {
    label: 'Genetic potential',
    url: '/genetic-potential',
  },
}
