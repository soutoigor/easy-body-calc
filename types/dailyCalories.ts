import { TBiologicalSex } from '@/types/biologicalSex'

export type TActivity = 1.2 | 1.375 | 1.55 | 1.725 | 1.9 | 0

export interface IBmr {
  weight: number
  height: number
  age: number
}

export interface IDailyCalories extends IBmr {
  biologicalSex: TBiologicalSex
  activity: TActivity
}

export interface IDailyCaloriesResult {
  bmr: number
  calories: number
}
