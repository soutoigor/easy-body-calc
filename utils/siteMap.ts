const siteMap = {
  BMI: {
    label: 'BMI rate',
    url: '/bmi',
  },
  BODY_FAT: {
    label: 'Body fat',
    url: '/body-fat',
  },
  MACROS: {
    label: 'Daily macros',
    url: '/daily-macros',
  },
  CALORIES: {
    label: 'Daily calories',
    url: '/bmr-daily-calories',
  },
}

export { siteMap }
