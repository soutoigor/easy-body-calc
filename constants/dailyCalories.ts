import { IActivities } from '@/types/dailyCalories'

export const ACTIVITY: IActivities = {
  SEDENTARY: 1.2,
  SLIGHTLY: 1.375,
  MODERATELY: 1.55,
  ACTIVE: 1.725,
  VERY: 1.9,
}
