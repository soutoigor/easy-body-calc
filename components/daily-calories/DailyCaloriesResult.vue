<template>
  <div class="result">
    <span class="result__text">Daily calories:</span>
    <span class="result__value">
      {{ calcResult.calories }}
    </span>
    <hr class="border-white border-opacity-50 w-3/5" />
    <span class="result__text">BMR:</span>
    <span class="result__value">
      {{ calcResult.bmr }}
    </span>
  </div>
</template>

<script lang="ts">
import { defineComponent, PropType } from '@nuxtjs/composition-api'
import { IDailyCaloriesResult } from '@/types/dailyCalories'

export default defineComponent({
  props: {
    calcResult: {
      type: Object as PropType<IDailyCaloriesResult>,
      required: true,
    },
  },
})
</script>

<style scoped>
.result {
  @apply border border-turbo rounded-lg h-80 w-full;
  @apply gap-5 flex flex-col items-center justify-center;
}

.result__value {
  @apply text-6xl italic font-black text-turbo;
}

.result__text {
  @apply italic capitalize font-bold text-2xl lg:text-3xl;
}
</style>
