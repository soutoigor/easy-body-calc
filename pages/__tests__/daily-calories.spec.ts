import { mount, RouterLinkStub } from '@vue/test-utils'
import Vue from 'vue'
import DailyCalories from '@/pages/bmr-daily-calories.vue'
import DailyCaloriesForm from '@/components/daily-calories/DailyCaloriesForm.vue'
import DailyCaloriesResult from '@/components/daily-calories/DailyCaloriesResult.vue'
import AppExplanation from '@/components/common/AppExplanation.vue'
import { IDailyCaloriesResult } from '@/types/dailyCalories'
import { SITE_MAP } from '@/constants/siteMap'

const WrapperComponent = Vue.component('WrapperComponent', {
  template: '<div><slot /></div>',
})

const mountComponent = () =>
  mount(WrapperComponent, {
    stubs: {
      NuxtLink: RouterLinkStub,
    },
    slots: {
      default: DailyCalories,
    },
  })

const result: IDailyCaloriesResult = {
  bmr: 1750,
  calories: 2000,
}

describe('DailyCalories', () => {
  it('should mount the component', () => {
    const wrapper = mountComponent()

    expect(wrapper.vm).toBeDefined()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should show form and hide the result when there is no "result"', () => {
    const wrapper = mountComponent()

    expect(wrapper.findComponent(AppExplanation).isVisible()).toBe(true)
    expect(wrapper.findComponent(DailyCaloriesForm).isVisible()).toBe(true)
    expect(wrapper.findComponent(DailyCaloriesResult).exists()).toBe(false)
  })

  it('should show result and hide form when there is "result"', async () => {
    const wrapper = mountComponent()

    wrapper.findComponent(DailyCaloriesForm).vm.$emit('result', result)

    await wrapper.vm.$nextTick()

    expect(wrapper.findComponent(DailyCaloriesResult).isVisible()).toBe(true)
    expect(wrapper.findComponent(AppExplanation).exists()).toBe(true)
    expect(wrapper.findComponent(DailyCaloriesForm).exists()).toBe(false)
  })

  it('should show "formTitle" according if there is "result"', async () => {
    const wrapper = mountComponent()

    const title = wrapper.find('.calc-section__title')

    expect(title.text()).toContain('Calculate your Daily Calories bellow:')

    wrapper.findComponent(DailyCaloriesForm).vm.$emit('result', result)

    await wrapper.vm.$nextTick()

    expect(title.text()).toContain('Your result:')
  })

  it('should display form again if "calculate again" button is clicked', async () => {
    const wrapper = mountComponent()

    wrapper.findComponent(DailyCaloriesForm).vm.$emit('result', result)

    await wrapper.vm.$nextTick()

    await wrapper.find('[data-testid="calculate-again"]').trigger('click')

    await wrapper.vm.$nextTick()

    expect(wrapper.findComponent(DailyCaloriesForm).isVisible()).toBe(true)
    expect(wrapper.findComponent(AppExplanation).isVisible()).toBe(true)
    expect(wrapper.findComponent(DailyCaloriesResult).exists()).toBe(false)
  })

  it('should call "router.push" with "macros" url and result calories as query param', async () => {
    const wrapper = mountComponent()

    wrapper.findComponent(DailyCaloriesForm).vm.$emit('result', result)

    await wrapper.vm.$nextTick()

    expect(wrapper.findComponent(RouterLinkStub).props().to).toStrictEqual({
      path: SITE_MAP.MACROS.url,
      query: { calories: result.calories.toString() },
    })
  })
})
