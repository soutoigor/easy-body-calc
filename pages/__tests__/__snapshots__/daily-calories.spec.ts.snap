// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DailyCalories should mount the component 1`] = `
"<div>
  <section class=\\"calc-section\\">
    <h2 class=\\"calc-section__title\\">Calculate your Daily Calories bellow:</h2>
    <transition-stub name=\\"rotate\\">
      <div class=\\"calc-section__form-container\\">
        <form class=\\"bodyfat-form\\">
          <div data-testid=\\"height\\"><label for=\\"input\\" class=\\"field__label\\">Height (cm)</label> <input id=\\"input\\" type=\\"text\\" autocomplete=\\"off\\" data-testid=\\"height\\" class=\\"field__input\\"></div>
          <div data-testid=\\"weight\\"><label for=\\"input\\" class=\\"field__label\\">Weight (kg)</label> <input id=\\"input\\" type=\\"text\\" autocomplete=\\"off\\" data-testid=\\"weight\\" class=\\"field__input\\"></div>
          <div data-testid=\\"age\\"><label for=\\"input\\" class=\\"field__label\\">Age</label> <input id=\\"input\\" type=\\"text\\" autocomplete=\\"off\\" data-testid=\\"age\\" class=\\"field__input\\"></div>
          <div><label for=\\"select\\" class=\\"field__label\\">Biological sex</label> <select id=\\"select\\" data-testid=\\"field\\" class=\\"field__input field__input--active\\">
              <option value=\\"M\\">
                Male
              </option>
              <option value=\\"F\\">
                Female
              </option>
            </select></div>
          <div><label for=\\"select\\" class=\\"field__label\\">Exercise level</label> <select id=\\"select\\" data-testid=\\"field\\" class=\\"field__input\\">
              <option value=\\"1.2\\">
                Sedentary
              </option>
              <option value=\\"1.375\\">
                Slightly active
              </option>
              <option value=\\"1.55\\">
                Moderately active
              </option>
              <option value=\\"1.725\\">
                Active
              </option>
              <option value=\\"1.9\\">
                Very active
              </option>
            </select></div> <button type=\\"submit\\" disabled=\\"disabled\\" class=\\"button primary\\" data-testid=\\"submit\\">
            Calculate
          </button>
        </form>
        <div><button type=\\"button\\" class=\\"button primary text\\" data-testid=\\"open-modal\\">
            How to get the measures
          </button>
          <transition-stub name=\\"modal\\" style=\\"display: none;\\">
            <div data-testid=\\"backdrop\\" class=\\"modal__backdrop\\">
              <article role=\\"dialog\\" aria-labelledby=\\"modalTitle\\" aria-describedby=\\"modalDescription\\" class=\\"modal__container\\">
                <header class=\\"modal__header\\">
                  <h1 class=\\"modal__title\\">How to get the measures</h1> <button data-testid=\\"closeButton\\" class=\\"modal__close\\"><svg xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 20 20\\" fill=\\"currentColor\\" class=\\"modal__close-icon\\">
                      <path fill-rule=\\"evenodd\\" d=\\"M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z\\" clip-rule=\\"evenodd\\"></path>
                    </svg></button>
                </header>
                <main class=\\"modal__content\\">
                  <div data-testid=\\"collapse\\">
                    <section class=\\"collapse\\">
                      <header class=\\"collapse__header\\">
                        <h1>Height</h1> <svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"collapse__icon\\">
                          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M19 9l-7 7-7-7\\"></path>
                        </svg>
                      </header>
                      <transition-stub name=\\"expand\\">
                        <!---->
                      </transition-stub>
                    </section>
                    <section class=\\"collapse\\">
                      <header class=\\"collapse__header\\">
                        <h1>Weight</h1> <svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"collapse__icon\\">
                          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M19 9l-7 7-7-7\\"></path>
                        </svg>
                      </header>
                      <transition-stub name=\\"expand\\">
                        <!---->
                      </transition-stub>
                    </section>
                    <section class=\\"collapse\\">
                      <header class=\\"collapse__header\\">
                        <h1>Hip</h1> <svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"collapse__icon\\">
                          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M19 9l-7 7-7-7\\"></path>
                        </svg>
                      </header>
                      <transition-stub name=\\"expand\\">
                        <!---->
                      </transition-stub>
                    </section>
                    <section class=\\"collapse\\">
                      <header class=\\"collapse__header\\">
                        <h1>Neck</h1> <svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"collapse__icon\\">
                          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M19 9l-7 7-7-7\\"></path>
                        </svg>
                      </header>
                      <transition-stub name=\\"expand\\">
                        <!---->
                      </transition-stub>
                    </section>
                    <section class=\\"collapse\\">
                      <header class=\\"collapse__header\\">
                        <h1>Abdomen</h1> <svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"collapse__icon\\">
                          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M19 9l-7 7-7-7\\"></path>
                        </svg>
                      </header>
                      <transition-stub name=\\"expand\\">
                        <!---->
                      </transition-stub>
                    </section>
                    <section class=\\"collapse\\">
                      <header class=\\"collapse__header\\">
                        <h1>Daily activities</h1> <svg xmlns=\\"http://www.w3.org/2000/svg\\" fill=\\"none\\" viewBox=\\"0 0 24 24\\" stroke=\\"currentColor\\" class=\\"collapse__icon\\">
                          <path stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" stroke-width=\\"2\\" d=\\"M19 9l-7 7-7-7\\"></path>
                        </svg>
                      </header>
                      <transition-stub name=\\"expand\\">
                        <!---->
                      </transition-stub>
                    </section>
                  </div>
                </main>
              </article>
            </div>
          </transition-stub>
        </div>
      </div>
    </transition-stub>
  </section>
  <aside class=\\"explanation-section\\">
    <article>
      <h1 class=\\"explanation-title\\">DAILY CALORIES</h1>
      <div class=\\"explanation-content\\">
        <p>
          Your <strong>daily calories</strong> is the sum of your
          <strong> basal metabolic rate</strong> (BMR) is the calories that
          your body needs to perform the body's most basic functions, like
          breathing, your brainc, etc, and the <strong>daily calories</strong>
          that your body use in your daily activities.
        </p>
        <p>
          This value represents the amount of calories that your body consumes
          during all the day. So, if you consume <b>more</b> calories than
          this result, you will gain <b>weight</b>, and consuming less, you
          will <strong>lose weight</strong>.
        </p>
      </div>
    </article>
  </aside>
</div>"
`;
